packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -462,6 +469,20 @@ export function renderWithHooks<Props, SecondArg>(
 
       workInProgress.updateQueue = null;
 
+      if (memoCache !== null) {
+        // Re-clone the cache, setting state in the middle of rendering can leave the cache
+        // in an invalid state
+        const previous = memoCache.previous;
+        if (previous !== null) {
+          memoCache.current = [...previous];
```

Comment:
Not sure how babel is configured, but `...` might test for an iterator, directly going to `slice()` is probably faster.

```suggestion
          memoCache.current = previous.slice();
```